### PR TITLE
Change peak finding algorithm

### PR DESCRIPTION
**Pull request by JPorron**
JPorron Large: 262 [](https://gitlab.cern.ch/rest-for-physics/rawlib/-/commits/jporron-work-on-PeakFinder) [](https://github.com/rest-for-physics/rawlib/commits/jporron-work-on-PeakFinder) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=rest-for-physics&utm_campaign=badge_info)

DON´T MERGE YET. THE PROCESS IS NOT FINISHED!

The peak finding algorithm is changed. The previous algorith compared each bin value with its closest neighbours and if it was bigger it was assigned as peak, producing too many peaks:

image

The new algorithm smooths the signal with a moving window that changes the i-bin value to the average of its neighbours (the number of neighbours can be chosen in the process ("const UShort_t smoothingWindow = 10;", in this case 10 means 5 bins to each side, so a window of 11 counting the i-bin).

After smoothing the criteria to consider a peak is to compare the smoothed value of the i-bin with the smoothed values of its "distance" neighbours (where "distance" is an input of the GetPeaks function that allows just one peak in it) and if it is bigger than all of the neighbours (but 2 to reduce the exclusion criteria) it is saved as peak.

Then a gauss fit with 20 bins to each side is performed to find the integer value of the peak bin. For the new algorithm:

image

(There are 2 peaks on top, one for each signal. They are very similar)

For this example 7 peaks are found and the amplitude values correctly correspond to the value of the bin considered as peak position:

image
image

### CODE REVIEW

**Comment by JPorron** [comment]
It seems to work okay for TPC signals, but it does´t for VETO signals (they are thinner)

**Comment by JPorron** [comment]
Relaxing the condition:
```
            bool isPeak = true;
            int numGreaterEqual = 0;  // Counter for smoothed values greater or equal to the studied bin

            for (UShort_t j = i - distance; j <= i + distance; ++j) {
                if (j != i && smoothedValue <= smoothedValues[j]) {
                    numGreaterEqual++;
                    if (numGreaterEqual >
                        2) {  // If more than one smoothed value is greater or equal, it's not a peak
                        isPeak = false;
                        break;
                    }
                }
            }
```
is key as two consecutive bins will have the same smoothed value if the value of the bin that leaves the moving window is the same as the value for the bin that enters. If the smoothed values are equal and we impose numGreaterEqual>0 none will be added.

This can lead to a bad estimation of the actual position of the peak, for example in this case: 

image

the calculated bin as maximum is 278 (which is not the biggest), but this effect it will be corrected by gauss fitting around the peak (giving a much more realistic value, 282 in this case). The results for numGreaterEqual>2 seem to be good even for events with many signals, where one peak is detected for each signal in a very reasonable position:

image

**Comment by JPorron** [comment]
For veto signals we set numGreaterEqual>0, as given their pointy nature the peak position is clear. No gaussian fit is performed, this peak position´s bin is the one saved as observable.

In order to consider a peak the user´s input is necessary as given the nature of the signals no method (at least i couldn´t think of one) using the signals baseline and sigma was appropiate. For a signal with a very prominent peak and a not so clear one, for example, given the prominent peak the signal´s baselinesigma is big and usual algorithms to peak identification as "baseline+baselinesigma * 5" make the second clear but much less prominent peak be under the threshold (due to the 5 * baselinesigma). 

The solution was to let the user set a threshold as a constant value over the signal´s threshold. For example, choosing 50 (which is appropiate for this example´s data):

image

**Comment by JPorron** [comment]
At this point some the peak finding seems good. The process still needs to be worked on to offer more information as classifying the vetos in groups, for example, as is done in TRestRawVetoAnalysisProcess.

**Comment by JPorron** [comment]
After adding the veto signals removal option, I don´t see a reason to do the VetoGroups separation, so I propose to close the Pull Request here.

The signals removal is selected in the .rml file with the new removal parameters:

image

When they are set to "false" or not specified all the signals will remain after this process. For a example of a experimental event:

image

We can see the correct identification of the peaks and their placing in their corresponding window (given this example´s value for the window). As we can see there is only one window for tpc signals and 2 for veto signals and no signals where removed:

image

If the peakless vetos are removed (\<parameter name="removePeaklessVetos" value="true"/\>):

image

The peak finding is unchanged but now, only the veto signals with peaks survive. The very small peaks in the left are part of some of the surviving veto signals but too low to be considered as a peak given this example´s value for "thresholdOverBaseline".

image

This 11 signals are the 11 different signal Ids (8 for tpc and 3 for veto) in the peaksChannelIds observables. Finally, removing all veto signals (\<parameter name="removeAllVetos" value="true"/\>) will, again, not change the peak observables but from now on there will be no veto signals:

image

Now only the tpc signals survive:

image

Adding both parameters (remove peakless vetos and then remove all vetos) will ignore the peakless removal and remove all the veto signals. 

Also, removing peakless vetos doesn´t make any sense if the process is applied for tpc signals, as it won´t know which veto signals are peakless, it will produce an error if channelType is "tpc" and the removal of peakless vetos is "ON". The removal of all vetos is allowed (we may not care about veto peaks and want to remove the signals from now on), in which case adding the veto removal for the "tpc" signals will not calculate the veto peaks:

image

Only the "tpc" peaks are found as, even though the veto peaksFinder is "ON" the veto signals where previously removed:

image

**Review by lobis — APPROVED** [review]

File: inc/TRestRawPeaksFinderProcess.h [hunk 1]
```diff
@@ -45,7 +49,7 @@ class TRestRawPeaksFinderProcess : public TRestEventProcess {
     TRestRawPeaksFinderProcess() = default;
     ~TRestRawPeaksFinderProcess() = default;
 
-    ClassDefOverride(TRestRawPeaksFinderProcess, 3);
+    ClassDefOverride(TRestRawPeaksFinderProcess, 5);
```
Comment (lobis) [inline inc/TRestRawPeaksFinderProcess.h:52]:
```suggestion
    ClassDefOverride(TRestRawPeaksFinderProcess, 4);
```

**Review by lobis — APPROVED** [review]

**Pull request merged**